elements/includes/creative_commons.inc

```diff
@@ -72,269 +124,1089 @@ function xml_form_elements_creative_commons($element, &$form_state) {
+      }
+      elseif ($response->getName() == 'error') {
+        drupal_set_message(t("The Creative Commons API unexpectedly returned an error. Creating license URI manually."), 'warning');
+      }
+      $new_license_uri = xml_form_elements_creative_commons_get_uri_from_values($values);
+    }
+    // If we had already a URI in manual mode and it differs from the new one,
+    // reset to manual now; otherwise store the new license URI.
+    if ($manual && (!empty($values['license_fieldset']['license_uri']) && $values['license_fieldset']['license_uri'] !== $new_license_uri)) {
+      $values['license_fieldset']['license_type'] = 'manual';
+    }
+    else {
+      $values['license_fieldset']['license_uri'] = $new_license_uri;
+    }
+    // Reset version overriding in automatic mode, if _user_ requested it.
+    if (!$manual && $override_version) {
+      $values['license_fieldset']['override_license_version'] = '1';
+    }
+  }
+
+  // If there was an error, return current values.
+  if ($error) {
+    // If in manual mode, re-set mode.
+    if ($manual) {
+      $values['license_fieldset']['license_type'] = 'manual';
+    }
+    return array($cc_prefix . $none_err, $values);
+  }
+
+  // If in manual mode, or if the previous request was unsuccessful, or if a
+  // different version was specified, query the API for validation for
+  // non-empty URIs.
+  if ((($values['license_fieldset']['license_type'] == 'manual') || !isset($new_version) || ($version != '' && ((float) $version != (float) $new_version)))
+    && !empty($values['license_fieldset']['license_uri'])) {
+    // Get the license details for the specified license URI.
+    $response = xml_form_elements_creative_commons_query_api_details($values['license_fieldset']['license_uri']);
+    if ($response) {
+      if ($response->getName() == 'error') {
+        // The license URI is not recognized.
+        $error = TRUE;
+        if ($values['license_fieldset']['license_type'] == 'manual') {
+          // If only the jurisdiction was unknown, alert the user; otherwise
+          // suggest choosing the license automatically.
+          if (strpos($values['license_fieldset']['license_uri'], $new_license_uri) === 0) {
+            form_set_error($element['#name'] . '][license_fieldset][license_uri', t("CC license jurisdiction unknown or not applicable for this license."));
+          }
+          else {
+            form_set_error($element['#name'] . '][license_fieldset][license_uri', t("CC license URI seems to be invalid. Perhaps you would prefer to choose one automatically?"));
+          }
+        }
+        else {
+          form_set_error($error_element, t("CC license version does not seem to exist."));
+        }
+      }
+    }
+    else {
+      // We could not reach the API.
+      drupal_set_message(t("Could not reach Creative Commons API to validate your input."), 'warning');
```

Comment:
This `else` here, what happens if Creative Commons can not be reached? Do we allow a save anyway? Or do we trigger an $error = TRUE? To be sure we don't get bad data saved? Thoughts?

Reply:
@DiegoPino the behaviour thus far (and, therefore, maybe the expected behaviour) was to "fall back" on manually creating the URL when Creative Commons is unreachable. I would incline to keep this (thus allowing unvalidated URLs in this case), and let the user decide upon the warning what should be done. What are @bondjimbond's thoughts about that?

Reply:
Ok, convinced... then lets tell people that if API fails manual is still an option? Too much to ask? Probably just a few more words after .. "validate your input" ...

Reply:
@DiegoPino If API is unavailable, users can still use _automatic_ mode. However the versions will be wrong (in all likelyhood). Or did I not understand your point?